### PR TITLE
Classifier - TESS Light Curve Viewer - Introduce D3 multiple brushes for annotations

### PR DESCRIPTION
**Pull request by shaunanoordin**
## PR Overview
package: `lib-classifier`
Requires #378 

This PR introduces [D3 brushes](https://github.com/d3/d3-brush) into the Light Curve Viewer, with the intent of using these brush constructs as UI components to represent Annotations on the LCV.

The goals of this PR:
- [x] users can use D3 brushes to mark areas of interest on the LCV
  - [x] multiple brushes can be created on the Light Curve Viewer at a time
- [x] Brushes should be restricted to the X-dimension
- [x] Creating a new brush should add a corresponding Annotation.
  - Each brush should be linked to an Annotation (henceforth, these constructs will be called **Annotation-Brushes** )
- [x] Annotation-Brushes should respect the Zoom/Pan of the LCV

Things to check for:
- [x] Annotation-Brushes are cleared between new Subjects
- [ ] Annotation-Brushes react properly when switching Tasks/Steps (i.e. removed when not on a graph2dRangeX tasks, and restored
- [x] Annotation-Brushes respect resizing of the browser window

### Dev Notes

**Multiple D3 Brushes**
- NOTE: by default, D3 doesn't do well with multiple brushes. A solution can be found here - http://bl.ocks.org/ludwigschubert/0236fa8594c4b02711b2606a8f95f605 (Credit to Ludwig Schubert) 
- TL;DR:
  1. brushes need to be stored in an array to keep track of 'em.
  2. there's always ONE brush more in the array than there are annotations, because that's the "create a new brush" interface.
  3. the default invisible "overlay" created by these brushes need to be cancelled out, otherwise we can't click on corresponding "create a new brush" interface.

On my ToDo list, but not for this PR:
- Modify the D3 brush construct to have a custom ' ✖️Delete' button
- When an Annotation-Brush is moved, the corresponding Annotation data structure needs to be updated

### Status
WIP

### CODE REVIEW

**Comment by shaunanoordin** [comment]
Current progress:

screen shot 2019-01-23 at 10 16 36

- Multiple brushes ✔️
- Can only create brushes during Annotate mode + graph2dRangeX Task ✖️ 
- each new brush creates a new Annotation ✖️

**Comment by shaunanoordin** [comment]
## PR Update
- Annotation-Brushes now properly respect zoom
- Annotations(-Brushes) can now be modified after they're created

**Pro-tip!** if you use `brush.move()` (e.g. to reposition a brush area after the zoom changes), be aware that this will trigger `brush.on('start/brushed/end')`. So if you have `brush.move()` _inside_ a `brush.on('end')`, then you need to be careful not to trigger infinite recursions.

**Comment by shaunanoordin** [comment]
## PR Update
- Annotation Brushes now save Annotations for Classifications!
- Old (non-D3 brush) Annotation logic has now been ripped out.

This PR has FINALLY reached parity with #378 (except for _"Annotation-Brushes react properly when switching Tasks/Steps (i.e. removed when not on a graph2dRangeX tasks, and restored",_ which isn't too important for this 25 Jan deadline)

This means that the usage of brushes now _doesn't_ break anything that was already working in #378.
- The LCV won't let you annotate if you're not in Annotate mode, or on the wrong Task
- Users can place Annotations,
- and Annotations respect the zoom feature
- Annotations are submitted correctly when Classifications are sent.

I'm going to close feature development on this PR soon, so I can spin off a new PR dedicated to the "Delete Annotations" feature (which will require modifying the brush structure, blargh), and maybe fix that onTaskChange logic (which will require `loadAnnotationToBrushes`).

NOTE: that onTaskChange thing is related to a quirk where you can TRY to _modify_ annotations when in the wrong task.

### Status
PR is feature-complete (so you can test on `http://localhost:8080/?project=1862`), but now it's _cleanup time_ (so you can actually read the code changes that I made)

**Comment by rogerhutchings** [comment]
This is going to need a rebase

**Comment by shaunanoordin** [comment]
This branch now includes updates from #394 and #400 . Let... the... rebasing... COMMENCE!

**Comment by shaunanoordin** [comment]
I've rebased this branch (locally) to the latest master (25f57e5037e48ec4bf134607498891d7c1ed7e46) but I'm now encountering an issue where neither Subjects nor Tasks will render properly.

screen shot 2019-02-12 at 18 03 56
_lib-classifier will show the Classifier and fetch the Project resource, but stops short of querying the next workflow._

Sarah mentioned that is related to an issue with `auth` that she's currently fixing, so I'll wait to re-rebase on those updates before I push my branch to this PR. 👍

**Comment by srallen** [comment]
PR that is being waited for is #446

**Comment by shaunanoordin** [comment]
## PR Update
This PR has been rebased to the latest master (4f47ecfcb770f2084ac2484d9ede83baca9797f8)!

screen shot 2019-02-14 at 17 41 02
_Tested on http://localhost:8080/?project=1862 - users can add Annotations, edit Annotations, remove Annotations, and pan/zoom. Annotations are recorded properly in the submitted Classifications._

The only additional update I added to this rebase is fixing a bug where `d3.event.stopPropagation` florped because there are occasions when `d3.event` is undefined.

### Status
This branch of development (#385 + #394 + #400) is now complete, so I'm going to go ahead and merge the launch-ready build of the TESS Light Curve Viewer, unless there are additional comments. 👌 As always, thanks to everyone for their feedback!
